coaster/views/classview.py

```diff
@@ -0,0 +1,383 @@
+        :param options: Additional options for :meth:`~flask.Flask.add_url_rule`
+        """
+        setattr(cls, _name, route(rule, **options)(cls.__get_raw_attr(_name)))
+
+    @classmethod
+    def init_app(cls, app, callback=None):
+        """
+        Register views on an app. If :attr:`callback` is specified, it will
+        be called after ``app.``:meth:`~flask.Flask.add_url_rule`, with the same
+        parameters.
+        """
+        processed = set()
+        for base in cls.__mro__:
+            for name, attr in base.__dict__.items():
+                if name in processed:
+                    continue
+                processed.add(name)
+                if isinstance(attr, ViewDecorator):
+                    attr.__set_name__(base, name)  # Required for Python < 3.6
+                    attr.init_app(app, cls, callback=callback)
+
+
+def _modelview_view_decorator(f):
+    @wraps(f)
+    def inner(self, **kwargs):
+        return f(self)
+    return inner
+
+
+class ModelView(ClassView):
+    """
+    Base class for constructing views around a model. Functionality is provided
+    via mixin classes that must precede :class:`ModelView` in base class order.
+    Two mixins are provided: :class:`UrlForView` and :class:`InstanceLoader`.
+    Sample use::
+
+        @route('/doc/<document>')
+        class DocumentView(UrlForView, InstanceLoader, ModelView):
+            model = Document
+            route_model_map = {
+                'document': 'name'
+                }
+
+            @route('')
+            @render_with(json=True)
+            def view(self):
+                return self.obj.current_access()
+
+        DocumentView.init_app(app)
+
+    :class:`ModelView` makes one significant departure from :class:`ClassView`:
+    view handler methods no longer receive URL rule variables as keyword
+    parameters. They are placed at ``self.kwargs`` instead, as it is assumed
+    that the view handler method has no further use for them once
+    :meth:`loader` loads the instance.
+    """
+    __decorators__ = ClassView.__decorators__ + [_modelview_view_decorator]
+
+    #: The model that this view class represents, to be specified by subclasses.
+    model = None
+
+    #: A mapping of URL rule variables to attributes on the model. For example,
+    #: if the URL rule is ``/<parent>/<document>``, the attribute map can be::
+    #:
+    #:     route_model_map = {
+    #:         'document': 'name',
+    #:         'parent': 'parent.name',
+    #:         }
+    route_model_map = {}
+
+    def loader(self):  # pragma: no cover
+        """
+        Subclasses or mixin classes may override this method to provide a model
+        instance loader. The return value of this method will be placed at
+        ``self.obj``.
+
+        TODO: Consider allowing :meth:`loader` to place attributes on ``self``
+        by itself, to accommodate scenarios where multiple models need to be
+        loaded.
```

Comment:
This will affect Funnel's `CommentSpace` and `VoteSpace`, since there's no backward link from them to `Proposal`. The loader must load the `Proposal` and join with `CommentSpace`/`VoteSpace`. In this case, the space goes into `self.obj` while the proposal goes into `self.proposal`.